### PR TITLE
Feature: Allow NULL in rows for computing correlations and covariance

### PR DESCRIPTION
**Pull request by Swatisoni**
JIRA: MADLIB-1166

Additional Authors:
Orhan Kislal <okislal@pivotal.io>

- Replace NULL values with the mean of corresponding column.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/290/

File: src/ports/postgres/modules/stats/correlation.py_in [hunk 1]
```diff
@@ -180,31 +180,29 @@ def _populate_output_table(schema_madlib, source_table, output_table,
             function_name = "Correlation"
             agg_str = "{0}.correlation_agg(x, mean)".format(schema_madlib)
 
+        cols = ','.join(["coalesce(" + col + ", avg_"+col+") "
+                            for col in col_names])
+        avgs = ','.join(["avg(" + col + ") AS avg_" + col + ""
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/correlation.py_in]: IMO, easier to read the above two strings without the concatenations i.e. 
```
cols = ','.join(["coalesce({0}, avg_{0})".format(col) for col in col_names])
avgs = ','.join(["avg({0}) AS avg_{0}".format(col) for col in col_names])
```

File: src/ports/postgres/modules/stats/correlation.py_in [hunk 2]
```diff
@@ -180,31 +180,29 @@ def _populate_output_table(schema_madlib, source_table, output_table,
...
+                            for col in col_names])
+        avg_array = ','.join(["avg_" + col + " " for col in col_names])
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/correlation.py_in]: Can add the trailing space in the `','` of the join.

File: src/ports/postgres/modules/stats/correlation.sql_in [hunk 3]
```diff
@@ -207,8 +203,9 @@ Result:
 </pre>
 
 @par Notes
-Current implementation ignores a row that contains NULL entirely. This means
-any correlation in such a row (with NULLs) does not contribute to the final answer.
+
+WARNING: Rows with NULL values will not be ignored. Null values will be
+replaced by the mean of their respective columns.
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/correlation.sql_in]: 1. The `@par Notes` adds formatting. Don't need the warning tag. 
2. I suggest removing the "Rows with NULL values will not be ignored."

Comment (fmcquillan99) [reply src/ports/postgres/modules/stats/correlation.sql_in]: Here is a blog post on the dangers of mean imputation
https://www.theanalysisfactor.com/mean-imputation/

Can we point to some literature (and put in the user docs) that back up our plan to use mean imputation?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/292/

**Comment by iyerr3** [comment]
Also please change commit message to better indicate the intention of the commit. 
Something along the lines of 
"Correlation: Impute NULL values with mean"

File: src/ports/postgres/modules/stats/correlation.py_in [hunk 4]
```diff
@@ -165,8 +165,8 @@ def _populate_output_table(schema_madlib, source_table, output_table,
     with MinWarning("info" if verbose else "error"):
         start = time()
         col_len = len(col_names)
-        col_names_as_float_array = py_list_to_sql_string(col_names, "float8")
         col_names_as_text_array = py_list_to_sql_string(col_names, "varchar")
+        col_names_as_float_array = py_list_to_sql_string(col_names, "float8")
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/correlation.py_in]: Don't need this variable anymore

File: src/ports/postgres/modules/stats/correlation.sql_in [hunk 5]
```diff
@@ -207,8 +204,17 @@ Result:
 </pre>
 
 @par Notes
-Current implementation ignores a row that contains NULL entirely. This means
-any correlation in such a row (with NULLs) does not contribute to the final answer.
+
+Null values will be replaced by the mean of their respective columns (Mean imputation/substitution). Mean imputation is a method in which the missing value on a certain variable is replaced by the mean of the available cases. This method maintains the sample size and is easy to use, but the variability in the data is reduced, so the standard deviations and the variance estimates tend to be underestimated. Please refer to [1] and [2] for details.
```
Comment (iyerr3) [inline src/ports/postgres/modules/stats/correlation.sql_in]: Need to wrap this to within 80 char length

**Comment by fmcquillan99** [comment]
The user doc explanation/caution looks good on mean imputation, thanks for adding

LGTM

**Comment by orhankislal** [comment]
Thank you @iyerr3 and @fmcquillan99 for your comments.

**Pull request closed without merging**